### PR TITLE
[G-API] Add support for more types serialization

### PR DESCRIPTION
**Pull request by smirnov-alexey**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake
```
force_builders=Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

Xbuild_image:Custom=centos:7
Xbuildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

build_image:Custom=ubuntu-openvino-2020.3.0:16.04
build_image:Custom Win=openvino-2020.3.0
build_image:Custom Mac=openvino-2020.3.0

test_modules:Custom=gapi
test_modules:Custom Win=gapi
test_modules:Custom Mac=gapi

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/gapi/src/backends/common/serialization.cpp [hunk 1]
```diff
@@ -813,6 +845,16 @@ I::IStream& ByteMemoryInStream::operator>> (std::string& str) {
     }
     return *this;
 }
+I::IStream& ByteMemoryInStream::operator>> (std::map<std::string, std::string>& map_str) {
```
Comment (smirnov-alexey) [inline modules/gapi/src/backends/common/serialization.cpp:848]: Probably should be a template map in the future

**Comment by smirnov-alexey** [comment]
@dmatveev , @anna-khakimova , please, review

File: modules/gapi/test/gapi_array_tests.cpp [hunk 2]
```diff
@@ -264,7 +264,7 @@ TEST(GArray_VectorRef, Kind)
     EXPECT_EQ(cv::detail::OpaqueKind::CV_SIZE, v7.getKind());
 
     cv::detail::VectorRef v8(std::vector<std::string>{});
-    EXPECT_EQ(cv::detail::OpaqueKind::CV_UNKNOWN, v8.getKind());
+    EXPECT_EQ(cv::detail::OpaqueKind::CV_STRING, v8.getKind());
```
Comment (dmatveev) [inline modules/gapi/test/gapi_array_tests.cpp:267]: We still need to test the types we don't know result in `CV_UNKNOWN`. The easiest way to guarantee there is a type G-API doesn't now is to define this type right in this test.

File: modules/gapi/test/gapi_opaque_tests.cpp [hunk 3]
```diff
@@ -233,7 +233,7 @@ TEST(GOpaque_OpaqueRef, Kind)
     EXPECT_EQ(cv::detail::OpaqueKind::CV_SIZE, v7.getKind());
 
     cv::detail::OpaqueRef v8(std::string{});
-    EXPECT_EQ(cv::detail::OpaqueKind::CV_UNKNOWN, v8.getKind());
+    EXPECT_EQ(cv::detail::OpaqueKind::CV_STRING, v8.getKind());
```
Comment (dmatveev) [inline modules/gapi/test/gapi_opaque_tests.cpp:236]: Same thing here

File: modules/gapi/src/backends/common/serialization.cpp [hunk 4]
```diff
@@ -734,6 +746,13 @@ I::OStream& ByteMemoryOutStream::operator<< (const std::string &str) {
     for (auto c : str) *this << c;
     return *this;
 }
+I::OStream& ByteMemoryOutStream::operator<< (const std::map<std::string, std::string> &map_str) {
+    *this << static_cast<uint32_t>(map_str.size());
+    for (const auto& strs : map_str) {
+        *this << strs.first << strs.second;
+    }
+    return *this;
+}
```
Comment (dmatveev) [inline modules/gapi/src/backends/common/serialization.cpp:755]: Instead, I'd recommend to implement a generic version like this:  https://github.com/opencv/opencv/blob/4.4.0/modules/gapi/src/backends/common/serialization.hpp#L192

File: modules/gapi/src/backends/common/serialization.cpp (hunk 1 above)
Comment (dmatveev) [reply modules/gapi/src/backends/common/serialization.cpp:850]: Let's do it now

File: modules/gapi/src/backends/common/serialization.cpp (hunk 4 above)
Comment (smirnov-alexey) [reply modules/gapi/src/backends/common/serialization.cpp]: Yeah, I saw this one. Added generic version

File: modules/gapi/test/gapi_opaque_tests.cpp (hunk 3 above)
Comment (smirnov-alexey) [reply modules/gapi/test/gapi_opaque_tests.cpp:244]: Done

File: modules/gapi/test/gapi_array_tests.cpp (hunk 2 above)
Comment (smirnov-alexey) [reply modules/gapi/test/gapi_array_tests.cpp:275]: Right, added a check on unknown type/kind

File: modules/gapi/src/backends/common/serialization.cpp (hunk 1 above)
Comment (smirnov-alexey) [reply modules/gapi/src/backends/common/serialization.cpp]: Done

**Review by dmatveev — APPROVED** [review]
Thanks! 👍 approved

**Comment by smirnov-alexey** [comment]
@alalek can we merge this?

**Review by anna-khakimova — APPROVED** [review]

**Comment by anna-khakimova** [comment]
Looks good.

**Pull request merged**